README.md
# karma-testingbot-launcher

[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square)](https://github.com/karma-runner/karma-testingbot-launcher)
 [![npm version](https://img.shields.io/npm/v/karma-testingbot-launcher.svg?style=flat-square)](https://www.npmjs.com/package/karma-testingbot-launcher) [![npm downloads](https://img.shields.io/npm/dm/karma-testingbot-launcher.svg?style=flat-square)](https://www.npmjs.com/package/karma-testingbot-launcher)

[![Build Status](https://img.shields.io/travis/testingbot/karma-testingbot-launcher/master.svg?style=flat-square)](https://travis-ci.org/testingbot/karma-testingbot-launcher) [![Dependency Status](https://img.shields.io/david/karma-runner/karma-testingbot-launcher.svg?style=flat-square)](https://david-dm.org/karma-runner/karma-testingbot-launcher) [![devDependency Status](https://img.shields.io/david/dev/karma-runner/karma-testingbot-launcher.svg?style=flat-square)](https://david-dm.org/karma-runner/karma-testingbot-launcher#info=devDependencies)


> Run your tests on the [TestingBot](https://testingbot.com/) browser cloud!


## Installation

Install `karma-testingbot-launcher` as a `devDependency` in your package.json:

```bash
npm install karma-testingbot-launcher --save-dev
```

## Usage

This launcher is used to run your tests across many browsers and platforms on TestingBot. Typically this runner is used in a CI (Continuous Integration) system.

### Adding karma-testingbot-launcher to an existing Karma config

To use and configure this launcher, you will need to add two properties to your top-level Karma config, `testingbot` and `customLaunchers`, set the `browsers` array to use TestingBot browsers, and add the `testingbot` reporter.

The `testingbot` object defines global properties for each browser/platform while the `customLaunchers` object configures individual browsers. The `testingbot` reporter will send back the test success state to https://testingbot.com. Here is a sample Karma config to get the launcher running:

```js
module.exports = function(config) {
  // Check out https://testingbot.com/support/getting-started/browsers.html for all browser possibilities
  var customLaunchers = {
    TB_chrome: {
      base: 'TestingBot',
      browserName: 'chrome',
      platform: 'Windows 7',
      version: '35'
    },
    TB_firefox: {
      base: 'TestingBot',
      browserName: 'firefox',
      version: '30'
    }
  }

  config.set({

    // The rest of your karma config is here
    // ...
    testingbot: {
        testName: 'Web App Unit Tests'
    },
    customLaunchers: customLaunchers,
    browsers: Object.keys(customLaunchers),
    reporters: ['dots', 'testingbot'],
    singleRun: true
  })
}
```

**Note: this config assumes that `process.env.TB_KEY` and `process.env.TB_SECRET` are set.**

## `testingbot` config properties shared across all browsers

### apiKey
Type: `String`
Default: `process.env.TB_KEY`

Your TestingBot api key, you can sign up [here](https://testingbot.com/users/sign_up).

### apiSecret
Type: `String`
Default: `process.env.TB_SECRET`

Your TestingBot api secret which you will see on your [account page](https://testingbot.com/members).

### startConnect
Type: `Boolean`
Default: `true`

If `true`, TestingBot Tunnel will be started automatically. Set this to `false` if you are launching tests locally and want to start TestingBot manually.

### connectOptions
Type: `Object`
Default:
```js
{
  apiKey: 'apiKey',
  apiSecret: 'apiSecret',
  tunnelIdentifier: 'tunnelId' //optional
}
```

Options to send to TestingBot. Check [here](https://testingbot.com/support/other/test-options) for all available options.

### seleniumHost
Type: `String`
default: `hub.testingbot.com`

### seleniumPort
Type: `Number`
default: `80`

### build
Type: `String`
Default: *One of the following environment variables*:
`process.env.BUILD_NUMBER`
`process.env.BUILD_TAG`
`process.env.CI_BUILD_NUMBER`
`process.env.CI_BUILD_TAG`
`process.env.TRAVIS_BUILD_NUMBER`
`process.env.CIRCLE_BUILD_NUM`
`process.env.DRONE_BUILD_NUMBER`

ID of the build currently running. This should be set by your CI.

### name
Type: `String`
Default: `'Karma test'`

Name of the unit test group you are running.

### tags
Type: `Array of Strings`

Tags to use for filtering jobs in your TestingBot account.

### tunnelIdentifier
Type: `String`

TestingBot Tunnel can proxy multiple sessions, this is an id of a tunnel session.


### screenrecorder
Type: `Boolean`
Default: `false`

Set to `true` if you want to record a video of your Karma session.

### screenshots
Type: `Boolean`
Default: `true`

Set to `false` if you don't want to record screenshots.

### public
Type: `String`
Default: `null`

Control who can view job details.

## `customLaunchers` config properties

The `customLaunchers` object has browser names as keys and configs as values. Documented below are the different properties which you can configure for each browser/platform combo.

*Note: You can learn about the available browser/platform combos on the [TestingBot browsers page](https://testingbot.com/support/getting-started/browsers.html).*

### base
Type: `String`
Required: `true`

This defines the base configuration for the launcher. In this case it should always be `TestingBot` so that browsers can use the base TestingBot config defined at the root `testingbot` property.

### browserName
Type: `String`
Required: `true`

Name of the browser.

### version
Type: `String`
Default: Latest browser version for all browsers

Version of the browser to use.

### platform
Type: `String`
Default: `'Linux'` for Firefox/Chrome, `'Windows 7'` for IE/Safari

Name of platform to run browser on.

### deviceOrientation
Type: `String`
Default: `'portrait'`

Accepted values: `'portrait' || 'landscape'`

Set this string if your unit tests need to run on a particular mobile device orientation for Android Browser or iOS Safari.

## More Information

You can find more information on our website regarding [TestingBot Tunnel](https://testingbot.com/support/other/tunnel)  and [Karma](https://testingbot.com/support/getting-started/karma.html).

CHANGELOG.md
<a name="1.0.2"></a>
## [1.0.2](https://github.com/karma-runner/karma-testingbot-launcher/compare/v1.0.0...v1.0.2) (2018-04-18)





examples/test/test.js
describe('add', function () {
  it('should add two numbers and return the result', function () {
    expect(window.add(1, 2)).toBe(3)
  })
})

describe('subtract', function () {
  it('should subtract two numbers', function () {
    expect(window.subtract(2, 1)).toBe(1)
  })
})

describe('updateAppState', function () {
  console.log('Hello from my test')

  it('should push a new state into the browser history', function () {
    window.updateAppState({
      message: 'hi'
    })
    expect(window.history.state).toEqual({
      message: 'hi'
    })
  })
})

gruntfile.js
module.exports = function (grunt) {
  grunt.initConfig({
    pkgFile: 'package.json',
    'npm-contributors': {
      options: {
        commitMessage: 'chore: update contributors'
      }
    },
    conventionalChangelog: {
      release: {
        options: {
          changelogOpts: {
            preset: 'angular'
          }
        },
        src: 'CHANGELOG.md'
      }
    },
    conventionalGithubReleaser: {
      release: {
        options: {
          auth: {
            type: 'oauth',
            token: process.env.GH_TOKEN
          },
          changelogOpts: {
            preset: 'angular',
            relaseCount: 0
          }
        }
      }
    },
    bump: {
      options: {
        commitMessage: 'chore: release v%VERSION%',
        pushTo: 'upstream',
        commitFiles: [
          'package.json',
          'CHANGELOG.md'
        ]
      }
    },
    karma: {
      example: {
        configFile: 'examples/karma.conf-ci.js'
      }
    },
    eslint: {
      target: [
        'lib/*.js',
        'index.js',
        'gruntfile.js',
        'examples/**/*.js'
      ]
    }
  })

  require('load-grunt-tasks')(grunt)

  grunt.registerTask('test', ['karma'])
  grunt.registerTask('default', ['eslint', 'test'])

  grunt.registerTask('release', 'Bump the version and publish to NPM.', function (type) {
    grunt.task.run([
      'npm-contributors',
      'bump:' + (type || 'patch') + ':bump-only',
      'conventionalChangelog',
      'bump-commit',
      'conventionalGithubReleaser',
      'npm-publish'
    ])
  })
}

lib/testingbot_tunnel.js
var q = require('q')
var launchTestingBotTunnel = require('testingbot-tunnel-launcher')

var testingbotTunnel = function (emitter, logger) {
  var log = logger.create('launcher.testingbot')
  var alreadyRunningDefered
  var alreadyRunningProces

  this.start = function (connectOptions, done) {
    if (alreadyRunningDefered) {
      log.debug('TestingBot Tunnel is already running or starting')
      return alreadyRunningDefered.promise
    }

    process.on('SIGINT', function () {
      if (alreadyRunningProces) {
        alreadyRunningProces.close()
      }
    })

    alreadyRunningDefered = q.defer()
    launchTestingBotTunnel(connectOptions, function (err, p) {
      if (err) {
        return alreadyRunningDefered.reject(err)
      }
      alreadyRunningProces = p
      alreadyRunningDefered.resolve()
    })

    return alreadyRunningDefered.promise
  }

  emitter.on('exit', function (done) {
    if (alreadyRunningProces) {
      log.info('Shutting down TestingBot Tunnel')
      setTimeout(function () {
        // wait a couple of seconds to make sure all http requests are processed
        alreadyRunningProces.close(done)
        alreadyRunningProces = null
      }, 2000)
    } else {
      done()
    }
  })
}

module.exports = testingbotTunnel


examples/karma.conf-ci.js
module.exports = function (config) {
  var customLaunchers = {
    'TB_Chrome': {
      base: 'TestingBot',
      platform: 'VISTA',
      browserName: 'chrome',
      customData: {
        awesome: true
      }
    },
    'TB_Firefox': {
      base: 'TestingBot',
      platform: 'VISTA',
      browserName: 'firefox'
    },
    'TB_IE': {
      base: 'TestingBot',
      platform: 'VISTA',
      browserName: 'internet explorer'
    }
  }

  config.set({
    basePath: '',
    frameworks: ['jasmine'],
    files: [
      'src/*.js',
      'test/*.js'
    ],
    reporters: ['progress', 'testingbot'],
    port: 4040,
    colors: true,
    logLevel: config.LOG_DEBUG,
    testingbot: {
      testName: 'Karma and TestingBot demo',
      recordScreenshots: false,
      connectOptions: {
        verbose: true,
        'se-port': 4445,
        logfile: 'testingbot_tunnel.log'
      },
      public: 'public'
    },
    // Increase timeout in case connection in CI is slow
    captureTimeout: 120000,
    customLaunchers: customLaunchers,
    browsers: Object.keys(customLaunchers),
    singleRun: true,
    concurrency: 4 // number of concurrent tests
  })
}


lib/testingbot_reporter.js
var TestingBotApi = require('testingbot-api')

var TBReporter = function (logger, /* testingbot:jobMapping */ jobMapping) {
  var log = logger.create('reporter.testingbot')

  var pendingUpdates = 0
  var updatesFinished = function () {}

  this.adapters = []

  // We're only interested in the final results per browser
  this.onBrowserComplete = function (browser) {
    var result = browser.lastResult

    // browser.launchId was used until v0.10.2, but changed to just browser.id in v0.11.0
    var browserId = browser.launchId || browser.id

    if (result.disconnected) {
      log.error('✖ Test Disconnected')
    }

    if (result.error) {
      log.error('✖ Test Errored')
    }

    if (browserId in jobMapping) {
      var jobDetails = jobMapping[browserId]
      var tbApi = new TestingBotApi(jobDetails.credentials)

      // We record pass/fail status
      var payload = {
        'test[success]': !(result.failed || result.error || result.disconnected) ? '1' : '0'
      }

      if (result.error) {
        payload['test[status_message]'] = result.error.toString()
      }

      pendingUpdates++

      tbApi.updateTest(payload, jobDetails.jobId, function (err) {
        pendingUpdates--
        if (err) {
          log.error('Failed record pass/fail status: %s', err.error)
        }

        if (pendingUpdates === 0) {
          updatesFinished()
        }
      })
    }
  }

  // Wait until all updates have been pushed
  this.onExit = function (done) {
    if (pendingUpdates) {
      updatesFinished = done
    } else {
      done()
    }
  }
}

module.exports = TBReporter


lib/testingbot_launcher.js
var wd = require('wd')

function formatError (err) {
  return err.message + '\n' + (err.data ? '  ' + err.data : '')
}

function processConfig (helper, config, args) {
  config = config || {}
  args = args || {}

  var apiKey = args.apiKey || config.apiKey || process.env.TB_KEY || process.env.TESTINGBOT_KEY
  var apiSecret = args.apiSecret || config.apiSecret || process.env.TB_SECRET || process.env.TESTINGBOT_SECRET
  var startConnect = config.startConnect !== false
  var tunnelIdentifier = args.tunnelIdentifier || config.tunnelIdentifier

  if (startConnect && !tunnelIdentifier) {
    tunnelIdentifier = 'karma' + Math.round(new Date().getTime() / 1000)
  }

  var browserName = args.browserName +
        (args.version ? ' ' + args.version : '') +
        (args.platform ? ' (' + args.platform + ')' : '')

  var connectOptions = helper.merge(config.connectOptions, {
    apiKey: apiKey,
    apiSecret: apiSecret,
    'se-port': 4445,
    'tunnel-identifier': tunnelIdentifier
  })

  var seleniumHost = config.seleniumHost || 'hub.testingbot.com'
  var seleniumPort = config.seleniumPort || 80

  var build = process.env.BUILD_NUMBER ||
        process.env.BUILD_TAG ||
        process.env.CI_BUILD_NUMBER ||
        process.env.CI_BUILD_TAG ||
        process.env.TRAVIS_BUILD_NUMBER ||
        process.env.CIRCLE_BUILD_NUM ||
        process.env.DRONE_BUILD_NUMBER

  var defaults = {
    version: '',
    platform: 'ANY',
    tags: [],
    name: 'Karma Test',
    'tunnel-identifier': tunnelIdentifier,
    'screenrecorder': false,
    'screenshots': false,
    'device-orientation': null,
    'disable-popup-handler': true,
    build: build || null,
    public: null,
    customData: {}
  }

  var options = helper.merge(
    // Legacy
    config.options,
    defaults, {
      // Pull out all the properties from the config that
      // we are interested in
      name: config.testName,
      build: config.build,
      'screenrecorder': config.recordVideo,
      'screenshots': config.recordScreenshots,
      public: config.public,
      customData: config.customData
    }, {
      // Need to rename some properties from args
      name: args.testName,
      'screenrecorder': args.recordVideo,
      'screenshots': args.recordScreenshots
    }, args
  )

  return {
    options: options,
    connectOptions: connectOptions,
    browserName: browserName,
    apiKey: apiKey,
    apiSecret: apiSecret,
    startConnect: startConnect,
    seleniumHost: seleniumHost,
    seleniumPort: seleniumPort
  }
}

var TBLauncher = function (
  args, testingbotTunnel,
  /* config.testingbot */ config,
  logger, helper,
  baseLauncherDecorator, captureTimeoutLauncherDecorator, retryLauncherDecorator,
  /* testingbot:jobMapping */ jobMapping
) {
  var self = this

  baseLauncherDecorator(self)
  captureTimeoutLauncherDecorator(self)
  retryLauncherDecorator(self)

  var pConfig = processConfig(helper, config, args)
  var options = pConfig.options
  var connectOptions = pConfig.connectOptions
  var browserName = pConfig.browserName
  var apiKey = pConfig.apiKey
  var apiSecret = pConfig.apiSecret
  var startConnect = pConfig.startConnect
  var seleniumHost = pConfig.seleniumHost
  var seleniumPort = pConfig.seleniumPort

  var pendingCancellations = 0
  var sessionIsReady = false

  self.name = browserName + ' on TestingBot'

  var pendingHeartBeat

  var log = logger.create('launcher.testingbot')
  var driverLog = logger.create('wd')

  var driver = wd.promiseChainRemote(seleniumHost, seleniumPort, apiKey, apiSecret)

  driver.on('status', function (info) {
    driverLog.debug(info.cyan)
  })

  driver.on('command', function (eventType, command, response) {
    driverLog.debug(' > ' + eventType.cyan, command, (response || '').grey)
  })

  driver.on('http', function (meth, path, data) {
    driverLog.debug(' > ' + meth.magenta, path, (data || '').grey)
  })

  var heartbeat = function () {
    pendingHeartBeat = setTimeout(function () {
      log.debug('Heartbeat to TestingBot (%s) - fetching title', browserName)

      driver.title()
        .then(null, function (err) {
          log.error('Heartbeat to %s failed\n  %s', browserName, formatError(err))

          clearTimeout(pendingHeartBeat)
          return self._done('failure')
        })

      heartbeat()
    }, 60000)
  }

  var start = function (url) {
    driver
      .init(options)
      .then(function () {
        if (pendingCancellations > 0) {
          pendingCancellations--
          return
        }
        // Record the job details, so we can access it later with the reporter
        jobMapping[self.id] = {
          jobId: driver.sessionID,
          credentials: {
            api_key: apiKey,
            api_secret: apiSecret
          }
        }

        sessionIsReady = true

        log.info('%s session at https://testingbot.com/tests/%s', browserName, driver.sessionID)
        log.debug('WebDriver channel for %s instantiated, opening %s', browserName, url)

        return driver.get(url)
          .then(heartbeat, function (err) {
            log.error('Can not start %s\n  %s', browserName, formatError(err))
            return self._done('failure')
          })
      }, function (err) {
        if (pendingCancellations > 0) {
          pendingCancellations--
          return
        }

        log.error('Can not start %s\n  %s', browserName, formatError(err))
        return self._done('failure')
      })
      .done()
  }

  self.on('start', function (url) {
    if (pendingCancellations > 0) {
      pendingCancellations--
      return
    }

    if (startConnect) {
      testingbotTunnel.start(connectOptions)
        .then(function () {
          if (pendingCancellations > 0) {
            pendingCancellations--
            return
          }

          start(url)
        }, function (err) {
          pendingCancellations--
          log.error('Can not start %s\n  Failed to start TestingBot Tunnel:\n  %s', browserName, err.message)

          self._retryLimit = -1 // don't retry
          self._done('failure')
        })
    } else {
      start(url)
    }
  })

  self.on('kill', function (done) {
    var allDone = function () {
      self._done()
      done()
    }

    if (sessionIsReady) {
      if (pendingHeartBeat) {
        clearTimeout(pendingHeartBeat)
      }

      log.debug('Shutting down the %s driver', browserName)
      driver.quit().nodeify(allDone)
      sessionIsReady = false
    } else {
      pendingCancellations++
      process.nextTick(allDone)
    }
  })
}

module.exports = TBLauncher
